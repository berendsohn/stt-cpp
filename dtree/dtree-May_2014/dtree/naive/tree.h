// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
