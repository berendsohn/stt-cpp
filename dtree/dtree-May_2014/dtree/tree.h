// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
