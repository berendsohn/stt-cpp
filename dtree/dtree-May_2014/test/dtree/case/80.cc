#define DTREE_CONFIG EndTreeWithDesc<WithAncDescAggr<dtree::DpMax<double>, dtree::DwGreaterEqual_<double>, WithAncAggr<dtree::DpMin<double>, dtree::DwLess_<double>, WithAncValue<dtree::DpAdd<double>, WithAncDescAggr<dtree::Count<int>, NullPredicate_, Begin > > > > >
#include "test/dtree/tree.cc"
