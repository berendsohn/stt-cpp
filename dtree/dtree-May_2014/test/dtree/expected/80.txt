<<<
  1: vL-1R-7F0 -> 14
  2: vL-4R24F0
  3: vL10R4F1 -> 5
  4: vL0R-2F0
  5: vL12R-25F1 -> 7
  6: vL-7R11F1 -> 4
  7: vL-8R-14F1 -> 12
  8: vL5R-2F0 -> 4
  9: vL-7R-8F1 -> 13
  10: vL4R-10F1 -> 5
  11: vL-8R-1F0
  12: vL-35R-21F1 -> 4
  13: vL3R-8F1 -> 4
  14: vL-26R1F1 -> 3
  15: vL-2R-3F0 -> 10
>>>
<<<
  @2: a1 o0 vL-4R24F0 aL0R0 a1 o0 vL-4R24F0 aL0R0 aL0R0 oL-infR-inf

  @11: a1 o0 vL-8R-1F0 aL0R0 a1 o0 vL-8R-1F0 aL0R0 aL0R0 oL-infR-inf

    /4: a1 o4 vL19R35F1 aL0R0 a1 o4 vL19R35F1 aL0R0 aL0R0 oL5R11
     ->13: a2 o4 vL3R-8F1 aL0R-10 a2 o4 vL3R-8F1 aL0R-10 aL3R-8 oL5R11
        \9: a1 o0 vL0R-10F0 aL0R0 a1 o0 vL0R-10F0 aL0R0 aL0R0 oL-infR-inf
       '.8: a1 o2 vL5R-2F0 aL0R0 a1 o2 vL5R-2F0 aL0R0 aL5R-2 oL5R11
         '.6: a1 o1 vL-7R11F1 aL0R0 a1 o1 vL-7R11F1 aL0R0 aL-7R11 oL-7R11
  @12: a7 o6 vL-35R-21F1 aL-4R0 a7 o6 vL-35R-21F1 aL-4R0 aL25R47 oL5R11
    \7: a5 o2 vL7R27F0 aL-11R-18 a5 o2 vL7R27F0 aL-11R-18 aL18R20 oL4R-3
      \5: a4 o2 vL-11R20F0 aL0R-38 a4 o2 vL-11R20F0 aL0R-38 aL29R0 oL4R-3
          /10: a1 o0 vL6R-7F1 aL0R0 a1 o0 vL6R-7F1 aL0R0 aL0R0 oL-infR-inf
       ->15: a2 o2 vL-2R-3F0 aL0R-7 a2 o2 vL-2R-3F0 aL0R-7 aL4R-3 oL4R-3
            /3: a1 o0 vL3R36F0 aL0R0 a1 o0 vL3R36F0 aL0R0 aL0R0 oL-infR-inf
          /14: a2 o0 vL-25R8F1 aL0R0 a2 o0 vL-25R8F1 aL0R0 aL3R36 oL-infR-inf
        \1: a3 o0 vL18R-13F1 aL-25R0 a3 o0 vL18R-13F1 aL-25R0 aL11R11 oL-infR-inf
>>>
